libsolc/libsolc.h

```diff
@@ -57,20 +61,44 @@ char const* solidity_license() SOLC_NOEXCEPT;
 /// The pointer returned must not be freed by the caller.
 char const* solidity_version() SOLC_NOEXCEPT;
 
+/// Allocates a chunk of memory of @p _size bytes.
+///
+/// Use this function inside callbacks to allocate data that is to be passed to
+/// the compiler. You may use solidity_free() or solidity_reset() to free this
+/// memory again but it is not required as the compiler takes ownership for any
+/// data passed to it via callbacks.
+///
+/// This function will return NULL if the requested memory region could not be allocated.
+char* solidity_alloc(size_t _size) SOLC_NOEXCEPT;
+
+/// Resizes @p _data to @p _size bytes.
+//
+/// Important, this call will abort() in case of any invalid argument being passed to this call.
+void solidity_resize(char* _data, size_t _size) SOLC_NOEXCEPT;
+
+/// Explicitly frees the memory that was being allocated with solidity_alloc(size_t size);
+/// or returned by a call to solidity_compile().
+///
+/// Important, this call will abort() in case of any invalid argument being passed to this call.
+void solidity_free(char* _data) SOLC_NOEXCEPT;
+
 /// Takes a "Standard Input JSON" and an optional callback (can be set to null). Returns
 /// a "Standard Output JSON". Both are to be UTF-8 encoded.
 ///
 /// @param _input The input JSON to process.
-/// @param _readCallback The optional callback pointer. Can be NULL.
+/// @param _readCallback The optional callback pointer. Can be NULL, but if not NULL,
+///                      it can be called by the compiler to request additional input.
+///                      Please see the documentation of the type for details.
 /// @param _readContext An optional context pointer passed to _readCallback. Can be NULL.
 ///
-/// @returns A pointer to the result. The pointer returned must not be freed by the caller.
+/// @returns A pointer to the result. The pointer returned must be freed by the caller.
 char const* solidity_compile(char const* _input, CStyleReadFileCallback _readCallback, void* _readContext) SOLC_NOEXCEPT;
```

Comment:
```suggestion
char const* solidity_compile(char const* _input, CStyleReadFileCallback _readCallback, void* _readContext);
```
By the way: ``solidity_compile`` should not be ``noexcept``. You could pass a callback to it that throws - and that would just work all the way through ``solidity_compile`` and be catchable outside of it, *unless* we declare it ``noexcept``. And there is hardly any gain in having it be ``noexcept``, so I'd argue it shouldn't be ``noexcept``, just the same way as ``qsort`` in the standard C library isn't ``noexcept``.

Reply:
This is a C interface. Not exceptions should cross C boundaries in my opinion.

Reply:
Well, since C++ doesn't have a well-defined stable ABI, you'd likely use the C interface from C++ code, whenever not linking statically... which we don't support for now, but still. I'm not saying we should promote throwing exceptions out of the callback, but since there's no actual benefit from declaring this ``noexcept``, I'd argue we shouldn't - ``noexcept`` is a very strong promise and should only be made where there is actual gain to it (as e.g. for ``solidity_alloc`` and ``solidity_free``).

Reply:
Actually, I'm against removing the `noexcept` from C API functions, because there should be no C++ exception leaking outside into the non-C++ world. I'd instead recommend try-catch'ing inside `solidity_compile()` and return a `nullptr` instead.

If error analysis is required, (could be added later), I'd suggest to provide a function that returns the error string (such as errno/strerror()/GetLastError() or similar).